src/main/java/io/supertokens/jwt/JWTSigningKey.java

```diff
@@ -0,0 +1,247 @@
+            List<JWTSigningKeyInfo> keysFromStorage = noSQLStorage.getJWTSigningKeys_Transaction();
+
+            if (keysFromStorage == null) {
+                keys = new ArrayList<>();
+            } else {
+                keys = keysFromStorage;
+            }
+
+            return keys;
+        }
+
+        throw new QuitProgramException("Unsupported storage type detected");
+    }
+
+    /**
+     * Used to retrieve a key for JWT validation, for a given signing algorithm. If there are no keys in storage that match the given algorithm a new one is generated.
+     * @param algorithm The signing algorithm
+     * @return {@link JWTSigningKeyInfo} key
+     * @throws UnsupportedJWTSigningAlgorithmException If there is an error in the provided algorithm when getting/generating keys
+     * @throws StorageQueryException If there is an error interacting with the database
+     * @throws StorageTransactionLogicException If there is an error interacting with the database
+     * @throws DuplicateKeyIdException If there is an error setting generated keys in storage
+     */
+    public JWTSigningKeyInfo getOrCreateAndGetKeyForAlgorithm(SupportedAlgorithms algorithm)
+            throws UnsupportedJWTSigningAlgorithmException, StorageQueryException, StorageTransactionLogicException, DuplicateKeyIdException {
+        SessionStorage storage = StorageLayer.getSessionStorage(main);
+
+        if (storage.getType() == STORAGE_TYPE.SQL) {
+            SessionSQLStorage sqlStorage = (SessionSQLStorage) storage;
+
+            try {
+                return sqlStorage.startTransaction(con -> {
+                    JWTSigningKeyInfo keyInfo = null;
+
+                    List<JWTSigningKeyInfo> keysFromStorage = sqlStorage.getJWTSigningKeys_Transaction(con);
+
+                    // If there are no keys in storage, generate a new one
+                    if (keysFromStorage.isEmpty()) {
+                        try {
+                            keyInfo = generateKeyForAlgorithm(algorithm);
+                            sqlStorage.setJWTSigningKey_Transaction(con, keyInfo);
+                        } catch (NoSuchAlgorithmException | DuplicateKeyIdException | UnsupportedJWTSigningAlgorithmException e) {
+                            throw new StorageTransactionLogicException(e);
+                        }
+                    } else {
+                        // Loop through the keys and find the first one for the algorithm
+                        for (int i = 0; i < keysFromStorage.size(); i++) {
+                            JWTSigningKeyInfo currentKey = keysFromStorage.get(i);
+                            if (currentKey.algorithm.equalsIgnoreCase(algorithm.name())) {
+                                keyInfo = currentKey;
+                                break;
+                            }
+                        }
+
+                        // If no key was found create a new one
+                        if (keyInfo == null) {
```

Comment:
Refactor code to prevent duplication of logic for creation of key.

Reply:
Done, moved to independent functions for SQL and NOSQL separately

Reply:
No i meant not to make it into a function, but to structure the code in a way where you need to run this logic just once per transaction - this is cause we can handle the DuplicateKeyId exception properly, just once. It's not really a big deal as such.. but just easier to read the code in the way I am thinking (I feel)

Reply:
Done, im still leaving the logic in the individual method because I think it makes the larger method look a little cleaner but I can change that back if you prefer

Reply:
I prefer if it's not in a function.

Reply:
Done